% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/simulate.R
\name{simFS}
\alias{simFS}
\title{Simulation of sequencing data for a single full-sib family.}
\usage{
simFS(rVec_f, rVec_m = rVec_f, epsilon = 0, config, nInd, meanDepth,
  thres = NULL, noFam = 1, noChr = 1, NoDS = 1,
  formats = list(gusmap = F, onemap = F, lepmap = F, joinmap = F, crimap
  = F), rd_dist = "Neg_Binom", filename = "sim", direct = "./",
  seed1 = 1, seed2 = 1)
}
\arguments{
\item{rVec_f, rVec_m}{Numeric vector of true paternal and maternal
recombination fractions (in the interval [0,0.5]). If only a single number
is given, then recombination between adjacent loci are equal.}

\item{epsilon}{Numeric value of the sequencing error rate.}

\item{config}{Numeric vector of the segregation types for all the loci.}

\item{nInd}{Positive integer value. The number of individuals.}

\item{meanDepth}{Positive numeric value. The mean depth of the read depth
distribution.}

\item{thres}{Numerical Value. Threshold value for which genotype calls with
a depth less than the threshold are set to missing.}

\item{NoDS}{Number of data sets to be simulated.}

\item{formats}{A list object where the element corresponding to the software
name is TRUE will produce data in that format. Currently handles software
for GUSMap, CRI-MAP, JoinMap, OneMap and Lep-MAP2.}

\item{rd_dist}{String. The distribution for which the read depths are
simulated from. Currently, only negative binomial ("Neg_Binom") and Poisson
("Pois") are implemented.}

\item{filename}{String. Common filename for the data sets generated from the
simulation.}

\item{seed1}{Numeric value. Random seed used for the simulation of the
parental phase (or OPGP).}

\item{seed2}{Numeric value. Random seed used for the simulation of the data
sets.}
}
\value{
If at least one method in \code{formats} argument is specified to be
TRUE, a file containing the simulation parameters is written and files
containing the simulated data for each of the methods is written. The naming
of the files (if \code{filename="sim"}) is as follows;
\itemize{
\item Simulation parameters: sim[No]_info.txt
\item GUSMap Genotypes: sim[No]_genon_SEQ.txt
\item GUSMap Read depths: sim[No]_ref_SEQ.txt
\item GUSMap Read depths: sim[No]_alt_SEQ.txt
\item OneMap format: sim[No]_OneMap.txt
\item LepMap2 format: sim[No]_LepMap.txt
\item CriMap format: chr1_sim[No].gen
\item JoinMap format: sim[No]_JoinMap.loc
}
where
"No" is the number of the data set simulated (from 1 up to \code{NoDS}).

If none of the methods in \code{formats} argument is specified to be TRUE,
then a list containing the following elements will be returned;
\itemize{
\item genon: Matrix of genotypes for the simulated sequencing data.
\item ref: Matrix of the allele counts for the reference allele.
\item alt: Matrix of the allele counts for the alternate allele.
\item trueGeno: Matrix of the true genotypes calls simulated.
\item rVec_f: Vector of adjacent paternal recombination fractions.
\item rVec_m: Vector of adjacent maternal recombination fractions.
\item nInd: Number of individuals.
\item nSnps: Number of SNPs.
\item config: Vector of segregation types used in the simulation.
\item OPGP: Vector of OPGPs used in the simulation.
\item meanDepth: Mean of the distribution used to simulate the read depths.
\item rd_dist: Distribution from the which the read depths where simulated.
\item epsilon: Numeric value of the sequencing error rate used in the
simulation.
}
}
\description{
Simulate genotypes and sequencing depth for the progeny of a full-sib family
and write the data to various linkage mapping software formats.
}
\details{
The \code{simFS} function simulates sequencing data for the progeny of a
full-sib family according to the simulation parameters assuming no
interference. The processed for the simulation is as follows;
\itemize{
\item Inheritance vectors are simulated according to the true recombination
fractions, assuming no interference.
\item The inheritance vectors are converted to genotypes for the specified
OPGP vector (which is simulated based on the \code{config} vector).
\item Simulated read depths are generated by simulating realizations from
the read depth distribution. The alleles of the true genotype are then
sampled with equal probability and replacement until a sample size
corresponding to the read depth is obtained, where a sequencing error
(e.g. an reference allele called as alternate and vice versa) is simulated
to occur with probability \code{epsilon}. The mean read depth for each loci
is assumed to be equal.
}

If none of the methods in \code{formats} argument is specified to be TRUE,
then no files are created and the simulated data is returned to an object
(see examples below). Otherwise, datasets are created and written to files
in the correct format for the various programs specified.

The value for the \code{config} vector are 1=both-informative (ABxAB),
2=paternal-informative A (ABxAA), 3=paternal-informative B (ABxBB),
4=maternal-informative (AAxAB), 5=maternal-informative (BBxAB), 6 =
uninformative (AAxAA), 7 = uninformative (AAxBB), 8 = uninformative (BBxAA)
and 9 = uninformative (BBxBB).
}
\examples{

## simulate full sib family
config <- c(2,1,1,4,2,4,1,1,4,1,2,1)
set.seed(3)
F1data <- simFS(runif(11,0.01,0.05), config=config, nInd=50, meanDepth=5)
## to look at the simulated data and simulation parameters
str(F1data)

## simulate full sib family assuming recombination is equal between loci
config <- c(2,1,1,4,2,4,1,1,4,1,2,1)
F1data <- simFS(0.01, config=config, nInd=50, meanDepth=5)

## To write the simulated data to files (in GUSMap format only). 
config <- c(2,1,1,4,2,4,1,1,4,1,2,1)
simFS(0.01, config=config, nInd=50, meanDepth=5, formats=list(gusmap=TRUE), filename="simTest")

}
\author{
Timothy P. Bilton
}
